{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import dill \n",
    "dill.dump_session(\"chr20_CCR_jason.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dill\n",
    "dill.load_session(\"chr20_CCR_jason.db\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Jason Kunisaki - Quinlan Lab Rotation\n",
    "## Evaluate CCRs\n",
    "\n",
    "This notebook contains the script needed to recapitulate CCRs from the first iteration of CCRs [(see manuscript)](https://www.nature.com/articles/s41588-018-0294-6.epdf?author_access_token=N8RkVYcavtplcSSE2KJkNdRgN0jAjWel9jnR3ZoTv0OG-o3UTB17cpoBs8B6XMBCl-5E0ZvpOii0iPl_hRSMGjWfkG4em7gjy95eV4bkTb0AF-E_dj3obeJfaadTja3aj9hUh1xk_BIztWgJScFx8w%3D%3D) and develop new filter parameters to better define CCRs. These parameters include **VAF, SIFT/PolyPhen scores, and coverage**. Use [PathoScore](https://github.com/quinlan-lab/pathoscore) to determine the efficacy of CCRs as a classifier for clinvar pathogenic variants. SIFT and PolyPhen interpretations can be found in the [VEP annotation descriptions](https://uswest.ensembl.org/info/genome/variation/prediction/protein_function.html). VAF filter is based on the percentile of VAFs across all variants from exome and whole genome sequencing (percentiles are calculated independent of sequencing used).   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "wget https://storage.googleapis.com/gnomad-public/release/2.1.1/vcf/exomes/gnomad.exomes.r2.1.1.sites.20.vcf.bgz\n",
    "gunzip -c gnomad.exomes.r2.1.1.sites.20.vcf.bgz > gnomad.exomes.r2.1.1.sites.20.vcf\n",
    "\n",
    "# Get necessary files \n",
    "```\n",
    "Exome Sequencing Gnomad file: \n",
    "cd ~/git/quinlan_lab_rotation/reference/\n",
    "wget https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/exomes/gnomad.exomes.r2.0.2.sites.vcf.bgz\n",
    "gunzip -c gnomad.exomes.r2.0.2.sites.vcf.bgz > gnomad.exomes.r2.0.2.sites.vcf\n",
    "\n",
    "WGS Gnomad file: \n",
    "wget https://storage.googleapis.com/gnomad-public/release/2.0.2/vcf/genomes/gnomad.genomes.r2.0.2.sites.coding_only.chr1-22.vcf.bgz\n",
    "gunzip -c gnomad.genomes.r2.0.2.sites.coding_only.chr1-22.vcf.bgz > gnomad.genomes.r2.0.2.sites.coding_only.chr1-22.vcf\n",
    "    \n",
    "GFF/GTF file: \n",
    "wget -P data ftp://ftp.ensembl.org/pub/grch37/current/gff3/homo_sapiens/Homo_sapiens.GRCh37.87.gff3.gz\n",
    "tabix Homo_sapiens.GRCh37.87.gff3.gz\n",
    "\n",
    "Exome Sequencing: \n",
    "cd ~/git/quinlan_lab_rotation/reference/coverage/exome\n",
    "for i in {1..22} X Y; do wget https://storage.googleapis.com/gnomad-public/release/2.0.2/coverage/exomes/gnomad.exomes.r2.0.2.chr$i.coverage.txt.gz; done\n",
    "\n",
    "\n",
    "Whole Genome Sequencing: \n",
    "cd ~/git/quinlan_lab_rotation/reference/coverage/wgs/\n",
    "for i in {1..22} X Y; do wget https://storage.googleapis.com/gnomad-public/release/2.0.2/coverage/genomes/gnomad.genomes.r2.0.2.chr$i.coverage.txt.gz; done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File name: new_CCR_no_sift_polyphen_filter_no_vaf_filter.bed\n"
     ]
    }
   ],
   "source": [
    "import cyvcf2\n",
    "import sys\n",
    "import exter ## exter.py file has to be in the working directory \n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from scipy.stats import percentileofscore\n",
    "import subprocess\n",
    "import os\n",
    "import fnmatch\n",
    "import csv\n",
    "from IPython.display import SVG\n",
    "from sklearn import preprocessing\n",
    "\n",
    "## Define the gff and vcf file path \n",
    "gff_path = \"/Users/jasonkunisaki/git/quinlan_lab_rotation/reference/Homo_sapiens.GRCh37.82.chr20.gff3\" \n",
    "exome_vcf_path = \"/Users/jasonkunisaki/git/quinlan_lab_rotation/reference/gnomad.exomes.r2.0.2.sites.vcf\"\n",
    "wgs_vcf_path = \"/Users/jasonkunisaki/git/quinlan_lab_rotation/reference/gnomad.genomes.r2.0.2.sites.coding_only.chr1-22.vcf\"\n",
    "exome_coverage_dir = \"~/git/quinlan_lab_rotation/reference/coverage/exome/gnomad.exomes.r2.0.2.chr\"\n",
    "wgs_coverage_dir = \"~/git/quinlan_lab_rotation/reference/coverage/exome/gnomad.genomes.r2.0.2.chr\"\n",
    "wgs_vcf_path = exome_vcf_path\n",
    "wgs_coverage_dir = exome_coverage_dir\n",
    "\n",
    "## Define SIFT/PolyPhen filter parameters \n",
    "sift_polyphen_filter = False\n",
    "sift_cutoff = 0.05\n",
    "polyphen_cutoff = 0.908\n",
    "\n",
    "## Define VAF filter parameters \n",
    "vaf_filter = False\n",
    "vaf_cutoff = 50 ## Percentile to use \n",
    "\n",
    "## Assign the pathoscore key name\n",
    "if sift_polyphen_filter: \n",
    "    pathoscore_name = \"new_CCR_\" + \"sift_cutoff_\" + str(sift_cutoff) + \"_polyphen_cutoff_\" + str(polyphen_cutoff)\n",
    "if not sift_polyphen_filter: \n",
    "    pathoscore_name = \"new_CCR_no_sift_polyphen_filter\"\n",
    "if vaf_filter: \n",
    "    pathoscore_name = pathoscore_name + \"_vaf_cutoff_\" + str(vaf_cutoff) + \"th_percentile\"\n",
    "if not vaf_filter: \n",
    "    pathoscore_name = pathoscore_name + \"_no_vaf_filter\"\n",
    "print(\"File name: \" + pathoscore_name + \".bed\")\n",
    "\n",
    "## Assign test gene name \n",
    "test_key = \"KCNQ2\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function that defines functional variants; see [manuscript](https://www.nature.com/articles/s41588-018-0294-6.epdf?author_access_token=N8RkVYcavtplcSSE2KJkNdRgN0jAjWel9jnR3ZoTv0OG-o3UTB17cpoBs8B6XMBCl-5E0ZvpOii0iPl_hRSMGjWfkG4em7gjy95eV4bkTb0AF-E_dj3obeJfaadTja3aj9hUh1xk_BIztWgJScFx8w%3D%3D) for more details "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to identify functional mutations\n",
    "def isfunctional(csq):\n",
    "    return any(c in csq['Consequence'] for c in ('stop_gained', 'stop_lost', 'start_lost', \n",
    "                                                 'initiator_codon_variant', 'rare_amino_acid_variant', \n",
    "                                                 'missense_variant', 'protein_altering_variant', \n",
    "                                                 'frameshift_variant', 'inframe_insertion', 'inframe_deletion')) \\\n",
    "    or (('splice_donor_variant' in csq['Consequence'] or 'splice_acceptor_variant' in csq['Consequence']) \\\n",
    "        and 'coding_sequence_variant' in csq['Consequence'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function to read through a VCF file and obtain all variants. This step requires the [cyvcf2](https://github.com/brentp/cyvcf2/tree/master/cyvcf2) function to read and process the vcf file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [],
   "source": [
    "def processVCF(vcf_path, gff_path, sift_polyphen_filter, sift_cutoff, polyphen_cutoff, type=\"exome\"):\n",
    "    ## Read in the VCF file \n",
    "    vcf = cyvcf2.VCF(vcf_path)\n",
    "    \n",
    "    ## Get the VEP description fields into a 'list'\n",
    "    kcsq = vcf[\"vep\"][\"Description\"].split(\":\")[1].strip(' \"').split(\"|\")\n",
    "    \n",
    "    ## Read in the gff transcript file and subset calls from chr20\n",
    "    transcripts = exter.read_gff(gff_path)\n",
    "    transcripts20 = transcripts[\"20\"] # get chr20\n",
    "\n",
    "    ## Make an empty dictionary \n",
    "    local_by_gene = defaultdict(list)\n",
    "    \n",
    "    ## For each variant in the VCF file: \n",
    "    for variant in vcf:\n",
    "        if variant.CHROM != \"20\": continue ## Remove non-chr20 variants (for now)\n",
    "        \n",
    "        ## Remove variants that have \"None, PASS, SEGDUP, or LCR\" as the filter result\n",
    "        if not (variant.FILTER is None or variant.FILTER in [\"PASS\", \"SEGDUP\", \"LCR\"]): continue\n",
    "        \n",
    "        ## Get the variant 'INFO' field from the VCF file for each variant\n",
    "        info = variant.INFO\n",
    "        \n",
    "        ## Set the functional consequence of the variant to 'False'\n",
    "        any_functional = False\n",
    "        \n",
    "        ## Merge the kcsq description keys with \n",
    "        csqs = [dict(zip(kcsq, c.split(\"|\"))) for c in info['vep'].split(\",\")]\n",
    "\n",
    "        ## For all variant/consequences that are within protein_coding regions\n",
    "        for csq in (c for c in csqs if c['BIOTYPE'] == 'protein_coding'):\n",
    "            ## Ignore intronic variants \n",
    "            if csq['Feature'] == '' or csq['EXON'] == '': continue #or csq['cDNA_position'] == '': continue\n",
    "            \n",
    "            ## Check to see if the variant's consequence is functional (based on the mutation types listed above)\n",
    "            if not isfunctional(csq): continue\n",
    "            \n",
    "            ## Assigns True to the variant as functional if it is NOT intronic and is deemed as functional by the 'isfunctional' function\n",
    "            any_functional = True\n",
    "\n",
    "            ## Get the exonic positions of the variants\n",
    "            local = transcripts20.localize(variant.start, variant.end)\n",
    "\n",
    "            ## Get the VAF of the variant\n",
    "            vaf = variant.INFO.get('AF')\n",
    "\n",
    "            ## Get the SIFT/PolyPhen results (IS THIS RIGHT)\n",
    "            SIFT = csq[\"SIFT\"]\n",
    "            PolyPhen = csq[\"PolyPhen\"]\n",
    "\n",
    "            ## Get the reference allele \n",
    "            ref_allele = variant.REF\n",
    "\n",
    "            ## Get string for the mutation \n",
    "            mut = []\n",
    "            for alt in variant.ALT: \n",
    "                mut.append(ref_allele + \">\" + alt) \n",
    "\n",
    "            ## Perform SIFT and Polyphen filters to remove the least delterious/damaging variant(s)\n",
    "            ## Go through each pair of scores and check if the varaint has SIFT and PolyPhen scores         \n",
    "            if sift_polyphen_filter == True: \n",
    "                sift_score = sift_cutoff + 0.1\n",
    "                polyphen_score = polyphen_cutoff - 0.1\n",
    "                if re.findall(\"deleterious\", SIFT):\n",
    "                    sift_score = float(re.split('\\)', re.split('\\(', SIFT)[1])[0])\n",
    "                if re.findall(\"damaging\", PolyPhen):\n",
    "                    polyphen_score = float(re.split('\\)', re.split('\\(', PolyPhen)[1])[0])\n",
    "                ## SIFT score has to be below the sift_cutoff to pass \n",
    "                ## PolyPhen score has to be above the polyphen_cutoff to pass \n",
    "                if not sift_score < sift_cutoff and polyphen_score > polyphen_cutoff:\n",
    "                    continue\n",
    "            if sift_polyphen_filter == False: \n",
    "                sift_score = SIFT\n",
    "                polyphen_score = PolyPhen\n",
    "            \n",
    "            for l in local:\n",
    "                l[\"chr\"] = variant.CHROM ## Get the chromosome \n",
    "                l[\"chr_start\"] = variant.start\n",
    "                l[\"chr_stop\"] = variant.end\n",
    "                l[\"mutation\"] = mut\n",
    "                l[\"vaf\"] = vaf\n",
    "                l[\"SIFT\"] = sift_score\n",
    "                l[\"PolyPhen\"] = polyphen_score\n",
    "                l[\"type\"] = type\n",
    "                local_by_gene[l[\"gene\"]].append(l)\n",
    "            break\n",
    "    return(local_by_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "b'vep'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-335-368489ff91e3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m## Read through gnomad WGS vcf variant file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mwgs_local_by_gene\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocessVCF\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvcf_path\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mwgs_vcf_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgff_path\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgff_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msift_polyphen_filter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msift_polyphen_filter\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msift_cutoff\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msift_cutoff\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpolyphen_cutoff\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpolyphen_cutoff\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"whole genome sequencing\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-334-3338b67fe3d3>\u001b[0m in \u001b[0;36mprocessVCF\u001b[0;34m(vcf_path, gff_path, sift_polyphen_filter, sift_cutoff, polyphen_cutoff, type)\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;31m## Get the VEP description fields into a 'list'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0mkcsq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvcf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"vep\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Description\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\":\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m' \"'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"|\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;31m## Read in the gff transcript file and subset calls from chr20\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/site-packages/cyvcf2/cyvcf2.pyx\u001b[0m in \u001b[0;36mcyvcf2.cyvcf2.VCF.__getitem__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/site-packages/cyvcf2/cyvcf2.pyx\u001b[0m in \u001b[0;36mcyvcf2.cyvcf2.VCF.get_header_type\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: b'vep'"
     ]
    }
   ],
   "source": [
    "## Read through gnomad WGS vcf variant file \n",
    "wgs_local_by_gene = processVCF(vcf_path=wgs_vcf_path, gff_path=gff_path, sift_polyphen_filter=sift_polyphen_filter, sift_cutoff=sift_cutoff, polyphen_cutoff=polyphen_cutoff, type=\"whole genome sequencing\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene = KCNQ2, variant start: 62037999 variant end: 62038000sift: deleterious_low_confidence(0) polyphen: possibly_damaging(0.628)\n",
      "gene = KCNQ2, variant start: 62038002 variant end: 62038003sift: deleterious_low_confidence(0.01) polyphen: benign(0.392)\n",
      "gene = KCNQ2, variant start: 62038010 variant end: 62038011sift: deleterious_low_confidence(0.01) polyphen: possibly_damaging(0.891)\n",
      "gene = KCNQ2, variant start: 62038013 variant end: 62038014sift: tolerated_low_confidence(0.92) polyphen: benign(0.014)\n",
      "gene = KCNQ2, variant start: 62038021 variant end: 62038022sift: tolerated_low_confidence(0.34) polyphen: benign(0.109)\n",
      "gene = KCNQ2, variant start: 62038043 variant end: 62038044sift: deleterious(0.01) polyphen: benign(0.22)\n",
      "gene = KCNQ2, variant start: 62038045 variant end: 62038046sift: deleterious(0) polyphen: probably_damaging(0.956)\n",
      "gene = KCNQ2, variant start: 62038049 variant end: 62038050sift: tolerated(1) polyphen: benign(0.287)\n",
      "gene = KCNQ2, variant start: 62038055 variant end: 62038056sift: deleterious(0.01) polyphen: possibly_damaging(0.886)\n",
      "gene = KCNQ2, variant start: 62038060 variant end: 62038061sift: tolerated(1) polyphen: possibly_damaging(0.726)\n",
      "gene = KCNQ2, variant start: 62038061 variant end: 62038062sift: tolerated(0.33) polyphen: possibly_damaging(0.452)\n",
      "gene = KCNQ2, variant start: 62038067 variant end: 62038068sift: deleterious(0) polyphen: possibly_damaging(0.9)\n",
      "gene = KCNQ2, variant start: 62038067 variant end: 62038068sift: deleterious(0) polyphen: possibly_damaging(0.9)\n",
      "gene = KCNQ2, variant start: 62038068 variant end: 62038069sift: deleterious(0.01) polyphen: probably_damaging(0.955)\n",
      "gene = KCNQ2, variant start: 62038069 variant end: 62038070sift: deleterious(0.03) polyphen: possibly_damaging(0.559)\n",
      "gene = KCNQ2, variant start: 62038072 variant end: 62038073sift: deleterious(0) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038075 variant end: 62038076sift: deleterious(0) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62038108 variant end: 62038109sift: deleterious(0) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62038111 variant end: 62038112sift: deleterious(0) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62038120 variant end: 62038121sift: deleterious(0.03) polyphen: probably_damaging(0.962)\n",
      "gene = KCNQ2, variant start: 62038127 variant end: 62038128sift: tolerated(0.34) polyphen: benign(0.009)\n",
      "gene = KCNQ2, variant start: 62038130 variant end: 62038131sift: tolerated(0.14) polyphen: benign(0.182)\n",
      "gene = KCNQ2, variant start: 62038132 variant end: 62038133sift: tolerated(0.05) polyphen: benign(0.011)\n",
      "gene = KCNQ2, variant start: 62038145 variant end: 62038146sift: tolerated(0.7) polyphen: benign(0.106)\n",
      "gene = KCNQ2, variant start: 62038145 variant end: 62038146sift: tolerated(0.24) polyphen: possibly_damaging(0.489)\n",
      "gene = KCNQ2, variant start: 62038150 variant end: 62038151sift: deleterious(0.01) polyphen: benign(0.031)\n",
      "gene = KCNQ2, variant start: 62038150 variant end: 62038151sift: deleterious(0.01) polyphen: possibly_damaging(0.693)\n",
      "gene = KCNQ2, variant start: 62038151 variant end: 62038152sift: tolerated(0.05) polyphen: possibly_damaging(0.876)\n",
      "gene = KCNQ2, variant start: 62038151 variant end: 62038152sift: tolerated(0.13) polyphen: benign(0.032)\n",
      "gene = KCNQ2, variant start: 62038177 variant end: 62038178sift: deleterious(0.04) polyphen: benign(0.128)\n",
      "gene = KCNQ2, variant start: 62038187 variant end: 62038188sift: deleterious(0.02) polyphen: possibly_damaging(0.769)\n",
      "gene = KCNQ2, variant start: 62038189 variant end: 62038190sift: deleterious(0) polyphen: probably_damaging(0.913)\n",
      "gene = KCNQ2, variant start: 62038205 variant end: 62038206sift: deleterious(0) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038216 variant end: 62038217sift: deleterious(0.03) polyphen: benign(0.329)\n",
      "gene = KCNQ2, variant start: 62038217 variant end: 62038218sift: deleterious(0) polyphen: probably_damaging(0.987)\n",
      "gene = KCNQ2, variant start: 62038238 variant end: 62038239sift: deleterious(0.01) polyphen: possibly_damaging(0.733)\n",
      "gene = KCNQ2, variant start: 62038238 variant end: 62038239sift: deleterious(0.02) polyphen: possibly_damaging(0.611)\n",
      "gene = KCNQ2, variant start: 62038270 variant end: 62038271sift: deleterious(0.01) polyphen: possibly_damaging(0.611)\n",
      "gene = KCNQ2, variant start: 62038271 variant end: 62038272sift: deleterious(0) polyphen: probably_damaging(0.944)\n",
      "gene = KCNQ2, variant start: 62038276 variant end: 62038277sift: tolerated(0.67) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62038279 variant end: 62038280sift: tolerated(0.12) polyphen: benign(0.008)\n",
      "gene = KCNQ2, variant start: 62038279 variant end: 62038280sift: tolerated(0.23) polyphen: benign(0.003)\n",
      "gene = KCNQ2, variant start: 62038283 variant end: 62038284sift: deleterious_low_confidence(0.05) polyphen: benign(0.207)\n",
      "gene = KCNQ2, variant start: 62038285 variant end: 62038286sift: tolerated_low_confidence(0.32) polyphen: benign(0.014)\n",
      "gene = KCNQ2, variant start: 62038285 variant end: 62038286sift: tolerated_low_confidence(0.75) polyphen: benign(0.003)\n",
      "gene = KCNQ2, variant start: 62038286 variant end: 62038287sift: tolerated_low_confidence(0.76) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62038288 variant end: 62038289sift: tolerated_low_confidence(0.24) polyphen: benign(0.003)\n",
      "gene = KCNQ2, variant start: 62038294 variant end: 62038295sift: tolerated_low_confidence(1) polyphen: benign(0.269)\n",
      "gene = KCNQ2, variant start: 62038298 variant end: 62038299sift: tolerated_low_confidence(0.57) polyphen: benign(0.004)\n",
      "gene = KCNQ2, variant start: 62038300 variant end: 62038301sift: tolerated(0.33) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62038303 variant end: 62038304sift: tolerated(0.27) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62038315 variant end: 62038316sift: tolerated(0.19) polyphen: benign(0.375)\n",
      "gene = KCNQ2, variant start: 62038320 variant end: 62038321sift: tolerated(0.57) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62038334 variant end: 62038335sift: tolerated(0.4) polyphen: benign(0.054)\n",
      "gene = KCNQ2, variant start: 62038334 variant end: 62038335sift: tolerated(0.12) polyphen: benign(0.294)\n",
      "gene = KCNQ2, variant start: 62038336 variant end: 62038337sift: deleterious_low_confidence(0.01) polyphen: possibly_damaging(0.9)\n",
      "gene = KCNQ2, variant start: 62038338 variant end: 62038339sift: tolerated_low_confidence(0.31) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62038349 variant end: 62038350sift: tolerated(1) polyphen: benign(0.006)\n",
      "gene = KCNQ2, variant start: 62038351 variant end: 62038352sift: tolerated_low_confidence(0.09) polyphen: benign(0.008)\n",
      "gene = KCNQ2, variant start: 62038355 variant end: 62038356sift: tolerated_low_confidence(0.22) polyphen: possibly_damaging(0.572)\n",
      "gene = KCNQ2, variant start: 62038363 variant end: 62038364sift: tolerated(0.14) polyphen: possibly_damaging(0.491)\n",
      "gene = KCNQ2, variant start: 62038371 variant end: 62038372sift: tolerated(0.15) polyphen: possibly_damaging(0.884)\n",
      "gene = KCNQ2, variant start: 62038375 variant end: 62038376sift: tolerated(0.16) polyphen: benign(0.236)\n",
      "gene = KCNQ2, variant start: 62038377 variant end: 62038381sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62038393 variant end: 62038394sift: deleterious_low_confidence(0.01) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62038397 variant end: 62038398sift: tolerated_low_confidence(0.05) polyphen: probably_damaging(0.921)\n",
      "gene = KCNQ2, variant start: 62038406 variant end: 62038407sift: tolerated(0.37) polyphen: benign(0.03)\n",
      "gene = KCNQ2, variant start: 62038411 variant end: 62038412sift: tolerated(0.08) polyphen: benign(0.012)\n",
      "gene = KCNQ2, variant start: 62038421 variant end: 62038422sift: tolerated(0.05) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038421 variant end: 62038422sift: deleterious(0.01) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038436 variant end: 62038437sift: tolerated(0.29) polyphen: benign(0.006)\n",
      "gene = KCNQ2, variant start: 62038441 variant end: 62038442sift: tolerated(0.25) polyphen: benign(0.057)\n",
      "gene = KCNQ2, variant start: 62038441 variant end: 62038442sift: tolerated(0.26) polyphen: benign(0.003)\n",
      "gene = KCNQ2, variant start: 62038441 variant end: 62038457sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62038442 variant end: 62038443sift: deleterious(0.03) polyphen: benign(0.399)\n",
      "gene = KCNQ2, variant start: 62038493 variant end: 62038494sift: tolerated(0.07) polyphen: benign(0.087)\n",
      "gene = KCNQ2, variant start: 62038495 variant end: 62038496sift: tolerated(0.39) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62038496 variant end: 62038497sift: tolerated(0.53) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62038501 variant end: 62038502sift: deleterious(0.03) polyphen: possibly_damaging(0.533)\n",
      "gene = KCNQ2, variant start: 62038502 variant end: 62038503sift: tolerated(0.15) polyphen: benign(0.318)\n",
      "gene = KCNQ2, variant start: 62038507 variant end: 62038508sift: tolerated(0.05) polyphen: benign(0.294)\n",
      "gene = KCNQ2, variant start: 62038510 variant end: 62038511sift: tolerated(0.22) polyphen: benign(0.074)\n",
      "gene = KCNQ2, variant start: 62038510 variant end: 62038514sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62038520 variant end: 62038521sift: tolerated(0.05) polyphen: benign(0.07)\n",
      "gene = KCNQ2, variant start: 62038525 variant end: 62038526sift: deleterious(0.02) polyphen: possibly_damaging(0.543)\n",
      "gene = KCNQ2, variant start: 62038528 variant end: 62038529sift: tolerated(0.21) polyphen: benign(0.003)\n",
      "gene = KCNQ2, variant start: 62038534 variant end: 62038535sift: tolerated(0.11) polyphen: benign(0.053)\n",
      "gene = KCNQ2, variant start: 62038543 variant end: 62038544sift: tolerated(0.07) polyphen: benign(0.052)\n",
      "gene = KCNQ2, variant start: 62038543 variant end: 62038544sift: tolerated(0.11) polyphen: benign(0.045)\n",
      "gene = KCNQ2, variant start: 62038547 variant end: 62038548sift: tolerated(0.08) polyphen: possibly_damaging(0.617)\n",
      "gene = KCNQ2, variant start: 62038548 variant end: 62038549sift: tolerated(0.14) polyphen: benign(0.017)\n",
      "gene = KCNQ2, variant start: 62038550 variant end: 62038551sift: tolerated(0.24) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62038559 variant end: 62038560sift: tolerated(0.37) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62038565 variant end: 62038566sift: tolerated(0.48) polyphen: benign(0.053)\n",
      "gene = KCNQ2, variant start: 62038566 variant end: 62038567sift: tolerated(0.55) polyphen: benign(0.006)\n",
      "gene = KCNQ2, variant start: 62038579 variant end: 62038580sift: tolerated(0.15) polyphen: benign(0.127)\n",
      "gene = KCNQ2, variant start: 62038585 variant end: 62038586sift: tolerated(0.08) polyphen: possibly_damaging(0.556)\n",
      "gene = KCNQ2, variant start: 62038586 variant end: 62038587sift: deleterious(0) polyphen: probably_damaging(0.921)\n",
      "gene = KCNQ2, variant start: 62038590 variant end: 62038591sift: tolerated(0.4) polyphen: benign(0.017)\n",
      "gene = KCNQ2, variant start: 62038591 variant end: 62038592sift: deleterious(0.02) polyphen: benign(0.005)\n",
      "gene = KCNQ2, variant start: 62038592 variant end: 62038593sift: deleterious(0.01) polyphen: possibly_damaging(0.653)\n",
      "gene = KCNQ2, variant start: 62038597 variant end: 62038598sift: deleterious(0.03) polyphen: benign(0.418)\n",
      "gene = KCNQ2, variant start: 62038609 variant end: 62038610sift: tolerated(0.11) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038612 variant end: 62038613sift: deleterious(0) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038618 variant end: 62038619sift: tolerated(0.22) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038624 variant end: 62038625sift: tolerated(0.12) polyphen: possibly_damaging(0.863)\n",
      "gene = KCNQ2, variant start: 62038627 variant end: 62038628sift: tolerated(0.07) polyphen: possibly_damaging(0.786)\n",
      "gene = KCNQ2, variant start: 62038639 variant end: 62038640sift: tolerated(0.06) polyphen: probably_damaging(0.984)\n",
      "gene = KCNQ2, variant start: 62038642 variant end: 62038643sift: deleterious(0.01) polyphen: probably_damaging(0.978)\n",
      "gene = KCNQ2, variant start: 62038643 variant end: 62038644sift: deleterious(0.01) polyphen: probably_damaging(0.97)\n",
      "gene = KCNQ2, variant start: 62038646 variant end: 62038647sift: tolerated(0.07) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62038649 variant end: 62038650sift: tolerated(0.3) polyphen: possibly_damaging(0.889)\n",
      "gene = KCNQ2, variant start: 62038658 variant end: 62038659sift: deleterious(0.05) polyphen: benign(0.252)\n",
      "gene = KCNQ2, variant start: 62038660 variant end: 62038661sift: tolerated(0.15) polyphen: possibly_damaging(0.473)\n",
      "gene = KCNQ2, variant start: 62038660 variant end: 62038661sift: tolerated(0.23) polyphen: benign(0.032)\n",
      "gene = KCNQ2, variant start: 62038660 variant end: 62038661sift: tolerated(0.74) polyphen: benign(0.054)\n",
      "gene = KCNQ2, variant start: 62038663 variant end: 62038664sift: tolerated(0.21) polyphen: possibly_damaging(0.867)\n",
      "gene = KCNQ2, variant start: 62038669 variant end: 62038670sift: deleterious(0.03) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038669 variant end: 62038670sift: tolerated(0.1) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038673 variant end: 62038674sift: tolerated(0.23) polyphen: benign(0.124)\n",
      "gene = KCNQ2, variant start: 62038675 variant end: 62038676sift: tolerated(0.06) polyphen: possibly_damaging(0.627)\n",
      "gene = KCNQ2, variant start: 62038676 variant end: 62038677sift: deleterious(0.01) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62038678 variant end: 62038679sift: tolerated(0.05) polyphen: probably_damaging(0.939)\n",
      "gene = KCNQ2, variant start: 62038680 variant end: 62038681sift: tolerated(0.4) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62038681 variant end: 62038682sift: tolerated(0.34) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62038682 variant end: 62038683sift: tolerated(0.39) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62038688 variant end: 62038689sift: tolerated(0.21) polyphen: possibly_damaging(0.607)\n",
      "gene = KCNQ2, variant start: 62038694 variant end: 62038695sift: deleterious(0) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62038717 variant end: 62038718sift: deleterious(0) polyphen: probably_damaging(0.996)\n",
      "gene = KCNQ2, variant start: 62039778 variant end: 62039779sift: tolerated(0.05) polyphen: possibly_damaging(0.69)\n",
      "gene = KCNQ2, variant start: 62039782 variant end: 62039783sift: tolerated(0.1) polyphen: probably_damaging(0.964)\n",
      "gene = KCNQ2, variant start: 62039803 variant end: 62039804sift: deleterious(0.03) polyphen: benign(0.139)\n",
      "gene = KCNQ2, variant start: 62039818 variant end: 62039819sift: tolerated(0.13) polyphen: possibly_damaging(0.516)\n",
      "gene = KCNQ2, variant start: 62039820 variant end: 62039821sift: tolerated(0.18) polyphen: benign(0.074)\n",
      "gene = KCNQ2, variant start: 62039821 variant end: 62039822sift: tolerated(0.67) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62039824 variant end: 62039825sift: tolerated(0.16) polyphen: benign(0.378)\n",
      "gene = KCNQ2, variant start: 62039829 variant end: 62039830sift: tolerated(0.12) polyphen: benign(0.357)\n",
      "gene = KCNQ2, variant start: 62039838 variant end: 62039839sift: tolerated(0.15) polyphen: benign(0.088)\n",
      "gene = KCNQ2, variant start: 62039838 variant end: 62039839sift: tolerated(0.65) polyphen: benign(0.015)\n",
      "gene = KCNQ2, variant start: 62039842 variant end: 62039843sift: deleterious(0) polyphen: probably_damaging(0.954)\n",
      "gene = KCNQ2, variant start: 62039842 variant end: 62039843sift: tolerated(0.07) polyphen: possibly_damaging(0.735)\n",
      "gene = KCNQ2, variant start: 62039851 variant end: 62039852sift: deleterious(0.03) polyphen: benign(0.025)\n",
      "gene = KCNQ2, variant start: 62039853 variant end: 62039854sift: deleterious(0.01) polyphen: possibly_damaging(0.584)\n",
      "gene = KCNQ2, variant start: 62039854 variant end: 62039855sift: tolerated(0.16) polyphen: benign(0.394)\n",
      "gene = KCNQ2, variant start: 62039859 variant end: 62039860sift: tolerated(0.47) polyphen: benign(0.013)\n",
      "gene = KCNQ2, variant start: 62039860 variant end: 62039861sift: tolerated(0.43) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62039862 variant end: 62039863sift: tolerated(0.27) polyphen: possibly_damaging(0.522)\n",
      "gene = KCNQ2, variant start: 62039875 variant end: 62039876sift: tolerated(0.62) polyphen: benign(0.32)\n",
      "gene = KCNQ2, variant start: 62039875 variant end: 62039876sift: deleterious(0.02) polyphen: probably_damaging(0.969)\n",
      "gene = KCNQ2, variant start: 62043146 variant end: 62043147sift: tolerated(0.06) polyphen: benign(0.08)\n",
      "gene = KCNQ2, variant start: 62043146 variant end: 62043147sift: tolerated(0.21) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62043163 variant end: 62043164sift: tolerated(0.23) polyphen: benign(0.01)\n",
      "gene = KCNQ2, variant start: 62043172 variant end: 62043173sift: tolerated(0.24) polyphen: benign(0.266)\n",
      "gene = KCNQ2, variant start: 62043173 variant end: 62043174sift: tolerated(0.34) polyphen: benign(0.007)\n",
      "gene = KCNQ2, variant start: 62043175 variant end: 62043176sift: tolerated(0.68) polyphen: benign(0.003)\n",
      "gene = KCNQ2, variant start: 62043190 variant end: 62043191sift: deleterious(0.02) polyphen: benign(0.193)\n",
      "gene = KCNQ2, variant start: 62043193 variant end: 62043194sift: deleterious(0.01) polyphen: benign(0.257)\n",
      "gene = KCNQ2, variant start: 62043194 variant end: 62043195sift: deleterious(0.01) polyphen: possibly_damaging(0.631)\n",
      "gene = KCNQ2, variant start: 62043197 variant end: 62043198sift: deleterious(0.02) polyphen: benign(0.197)\n",
      "gene = KCNQ2, variant start: 62043211 variant end: 62043212sift: deleterious(0.01) polyphen: benign(0.085)\n",
      "gene = KCNQ2, variant start: 62043218 variant end: 62043219sift: deleterious(0.03) polyphen: benign(0.163)\n",
      "gene = KCNQ2, variant start: 62043223 variant end: 62043224sift: deleterious(0) polyphen: benign(0.207)\n",
      "gene = KCNQ2, variant start: 62043226 variant end: 62043227sift: deleterious(0.03) polyphen: benign(0.141)\n",
      "gene = KCNQ2, variant start: 62043226 variant end: 62043227sift: tolerated(0.06) polyphen: benign(0.054)\n",
      "gene = KCNQ2, variant start: 62043226 variant end: 62043227sift: deleterious(0.01) polyphen: benign(0.039)\n",
      "gene = KCNQ2, variant start: 62043228 variant end: 62043229sift: deleterious(0.02) polyphen: benign(0.01)\n",
      "gene = KCNQ2, variant start: 62043234 variant end: 62043235sift: deleterious(0.02) polyphen: benign(0.055)\n",
      "gene = KCNQ2, variant start: 62044763 variant end: 62044764sift: tolerated(0.05) polyphen: benign(0.341)\n",
      "gene = KCNQ2, variant start: 62044875 variant end: 62044876sift: deleterious(0) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62044908 variant end: 62044909sift: deleterious(0) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62044917 variant end: 62044918sift: tolerated(0.34) polyphen: benign(0.248)\n",
      "gene = KCNQ2, variant start: 62044925 variant end: 62044926sift: tolerated(0.45) polyphen: possibly_damaging(0.614)\n",
      "gene = KCNQ2, variant start: 62044932 variant end: 62044933sift: tolerated(1) polyphen: benign(0.028)\n",
      "gene = KCNQ2, variant start: 62045444 variant end: 62045445sift: deleterious(0.03) polyphen: possibly_damaging(0.9)\n",
      "gene = KCNQ2, variant start: 62045461 variant end: 62045462sift: deleterious(0.01) polyphen: benign(0.347)\n",
      "gene = KCNQ2, variant start: 62045470 variant end: 62045471sift: deleterious(0) polyphen: probably_damaging(0.964)\n",
      "gene = KCNQ2, variant start: 62045482 variant end: 62045483sift: deleterious(0) polyphen: benign(0.373)\n",
      "gene = KCNQ2, variant start: 62045483 variant end: 62045484sift: deleterious(0) polyphen: benign(0.388)\n",
      "gene = KCNQ2, variant start: 62045486 variant end: 62045487sift: tolerated(0.27) polyphen: benign(0.007)\n",
      "gene = KCNQ2, variant start: 62045492 variant end: 62045493sift: tolerated(0.14) polyphen: probably_damaging(0.991)\n",
      "gene = KCNQ2, variant start: 62045495 variant end: 62045496sift: deleterious(0) polyphen: probably_damaging(0.922)\n",
      "gene = KCNQ2, variant start: 62045498 variant end: 62045499sift: tolerated(0.14) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62045512 variant end: 62045513sift: deleterious(0) polyphen: benign(0.07)\n",
      "gene = KCNQ2, variant start: 62045513 variant end: 62045514sift: tolerated(0.19) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62045521 variant end: 62045522sift: tolerated(0.17) polyphen: benign(0.127)\n",
      "gene = KCNQ2, variant start: 62045522 variant end: 62045523sift: tolerated(0.74) polyphen: benign(0.004)\n",
      "gene = KCNQ2, variant start: 62045523 variant end: 62045524sift: tolerated(0.46) polyphen: benign(0.028)\n",
      "gene = KCNQ2, variant start: 62045526 variant end: 62045527sift: tolerated(0.43) polyphen: possibly_damaging(0.752)\n",
      "gene = KCNQ2, variant start: 62045526 variant end: 62045527sift: tolerated(0.43) polyphen: possibly_damaging(0.752)\n",
      "gene = KCNQ2, variant start: 62045531 variant end: 62045532sift: deleterious(0.01) polyphen: probably_damaging(0.968)\n",
      "gene = KCNQ2, variant start: 62046257 variant end: 62046258sift: deleterious(0) polyphen: benign(0.003)\n",
      "gene = KCNQ2, variant start: 62046263 variant end: 62046264sift: tolerated(0.75) polyphen: possibly_damaging(0.69)\n",
      "gene = KCNQ2, variant start: 62046264 variant end: 62046265sift: deleterious(0) polyphen: probably_damaging(0.967)\n",
      "gene = KCNQ2, variant start: 62046264 variant end: 62046265sift: deleterious(0.01) polyphen: probably_damaging(0.967)\n",
      "gene = KCNQ2, variant start: 62046269 variant end: 62046270sift: tolerated(0.13) polyphen: probably_damaging(0.987)\n",
      "gene = KCNQ2, variant start: 62046275 variant end: 62046276sift: deleterious(0.04) polyphen: benign(0.092)\n",
      "gene = KCNQ2, variant start: 62046276 variant end: 62046277sift: tolerated(0.31) polyphen: benign(0.008)\n",
      "gene = KCNQ2, variant start: 62046278 variant end: 62046279sift: tolerated(0.44) polyphen: benign(0.05)\n",
      "gene = KCNQ2, variant start: 62046284 variant end: 62046285sift: tolerated(0.22) polyphen: benign(0.004)\n",
      "gene = KCNQ2, variant start: 62046290 variant end: 62046291sift: deleterious(0.01) polyphen: possibly_damaging(0.888)\n",
      "gene = KCNQ2, variant start: 62046291 variant end: 62046292sift: deleterious(0.02) polyphen: benign(0.031)\n",
      "gene = KCNQ2, variant start: 62046294 variant end: 62046295sift: deleterious(0.01) polyphen: possibly_damaging(0.543)\n",
      "gene = KCNQ2, variant start: 62046297 variant end: 62046298sift: deleterious(0.01) polyphen: benign(0.149)\n",
      "gene = KCNQ2, variant start: 62046302 variant end: 62046303sift: deleterious(0) polyphen: benign(0.022)\n",
      "gene = KCNQ2, variant start: 62046302 variant end: 62046303sift: deleterious(0.01) polyphen: possibly_damaging(0.888)\n",
      "gene = KCNQ2, variant start: 62046303 variant end: 62046304sift: deleterious(0) polyphen: possibly_damaging(0.888)\n",
      "gene = KCNQ2, variant start: 62046309 variant end: 62046310sift: deleterious(0) polyphen: probably_damaging(0.944)\n",
      "gene = KCNQ2, variant start: 62046315 variant end: 62046316sift: deleterious(0) polyphen: possibly_damaging(0.888)\n",
      "gene = KCNQ2, variant start: 62046315 variant end: 62046316sift: deleterious(0.01) polyphen: benign(0.33)\n",
      "gene = KCNQ2, variant start: 62046321 variant end: 62046322sift: deleterious(0.02) polyphen: possibly_damaging(0.868)\n",
      "gene = KCNQ2, variant start: 62046360 variant end: 62046361sift: deleterious(0.02) polyphen: benign(0.007)\n",
      "gene = KCNQ2, variant start: 62046365 variant end: 62046366sift: tolerated(0.29) polyphen: benign(0.008)\n",
      "gene = KCNQ2, variant start: 62046372 variant end: 62046373sift: tolerated(0.21) polyphen: benign(0.013)\n",
      "gene = KCNQ2, variant start: 62046375 variant end: 62046376sift: tolerated(0.48) polyphen: benign(0.206)\n",
      "gene = KCNQ2, variant start: 62046387 variant end: 62046388sift: deleterious(0.01) polyphen: probably_damaging(0.946)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene = KCNQ2, variant start: 62046388 variant end: 62046389sift: tolerated(0.11) polyphen: benign(0.037)\n",
      "gene = KCNQ2, variant start: 62046393 variant end: 62046394sift: tolerated(0.37) polyphen: benign(0.037)\n",
      "gene = KCNQ2, variant start: 62046397 variant end: 62046398sift: deleterious(0.03) polyphen: possibly_damaging(0.847)\n",
      "gene = KCNQ2, variant start: 62046401 variant end: 62046402sift: tolerated(0.36) polyphen: benign(0.008)\n",
      "gene = KCNQ2, variant start: 62046402 variant end: 62046403sift: tolerated(0.17) polyphen: benign(0.149)\n",
      "gene = KCNQ2, variant start: 62046407 variant end: 62046408sift: deleterious(0.03) polyphen: probably_damaging(0.95)\n",
      "gene = KCNQ2, variant start: 62046408 variant end: 62046409sift: tolerated(0.28) polyphen: probably_damaging(0.923)\n",
      "gene = KCNQ2, variant start: 62046413 variant end: 62046414sift: tolerated(0.11) polyphen: possibly_damaging(0.571)\n",
      "gene = KCNQ2, variant start: 62046414 variant end: 62046415sift: deleterious(0.01) polyphen: benign(0.065)\n",
      "gene = KCNQ2, variant start: 62046428 variant end: 62046429sift: tolerated(0.21) polyphen: benign(0.011)\n",
      "gene = KCNQ2, variant start: 62046432 variant end: 62046433sift: tolerated(0.17) polyphen: benign(0.025)\n",
      "gene = KCNQ2, variant start: 62046435 variant end: 62046436sift: tolerated(0.12) polyphen: benign(0.026)\n",
      "gene = KCNQ2, variant start: 62046441 variant end: 62046442sift: deleterious(0.01) polyphen: possibly_damaging(0.615)\n",
      "gene = KCNQ2, variant start: 62046459 variant end: 62046460sift: deleterious(0.02) polyphen: benign(0.275)\n",
      "gene = KCNQ2, variant start: 62046467 variant end: 62046468sift: tolerated(0.06) polyphen: possibly_damaging(0.827)\n",
      "gene = KCNQ2, variant start: 62046471 variant end: 62046472sift: tolerated(0.16) polyphen: benign(0.034)\n",
      "gene = KCNQ2, variant start: 62050972 variant end: 62050973sift: deleterious(0.02) polyphen: benign(0.081)\n",
      "gene = KCNQ2, variant start: 62050977 variant end: 62050978sift: tolerated(0.24) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62050977 variant end: 62050978sift: tolerated(0.28) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62050978 variant end: 62050979sift: deleterious(0.04) polyphen: benign(0.118)\n",
      "gene = KCNQ2, variant start: 62050981 variant end: 62050982sift: tolerated(0.24) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62050983 variant end: 62050984sift: deleterious(0.04) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62050990 variant end: 62050991sift: tolerated(0.09) polyphen: benign(0.054)\n",
      "gene = KCNQ2, variant start: 62050997 variant end: 62051000sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62051002 variant end: 62051003sift: tolerated(0.59) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62051013 variant end: 62051014sift: tolerated(0.24) polyphen: benign(0.01)\n",
      "gene = KCNQ2, variant start: 62051016 variant end: 62051017sift: tolerated(0.42) polyphen: benign(0.017)\n",
      "gene = KCNQ2, variant start: 62051019 variant end: 62051020sift: tolerated(0.45) polyphen: benign(0.036)\n",
      "gene = KCNQ2, variant start: 62051019 variant end: 62051020sift: tolerated(0.5) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62055544 variant end: 62055545sift: deleterious(0.03) polyphen: benign(0.013)\n",
      "gene = KCNQ2, variant start: 62055546 variant end: 62055548sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62055547 variant end: 62055548sift: tolerated(0.72) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62055553 variant end: 62055554sift: tolerated(0.06) polyphen: benign(0.004)\n",
      "gene = KCNQ2, variant start: 62059720 variant end: 62059721sift: deleterious(0) polyphen: possibly_damaging(0.557)\n",
      "gene = KCNQ2, variant start: 62059726 variant end: 62059727sift: tolerated(1) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62059741 variant end: 62059742sift: tolerated(0.36) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62059749 variant end: 62059750sift: tolerated(0.33) polyphen: benign(0.093)\n",
      "gene = KCNQ2, variant start: 62059777 variant end: 62059778sift: tolerated(0.35) polyphen: possibly_damaging(0.783)\n",
      "gene = KCNQ2, variant start: 62059780 variant end: 62059781sift: tolerated(0.23) polyphen: probably_damaging(0.935)\n",
      "gene = KCNQ2, variant start: 62059780 variant end: 62059781sift: tolerated(0.05) polyphen: possibly_damaging(0.831)\n",
      "gene = KCNQ2, variant start: 62062702 variant end: 62062703sift: tolerated(0.25) polyphen: possibly_damaging(0.58)\n",
      "gene = KCNQ2, variant start: 62062704 variant end: 62062705sift: tolerated(0.47) polyphen: benign(0.011)\n",
      "gene = KCNQ2, variant start: 62062707 variant end: 62062708sift: tolerated(0.71) polyphen: benign(0.007)\n",
      "gene = KCNQ2, variant start: 62062717 variant end: 62062718sift: tolerated(0.86) polyphen: benign(0.031)\n",
      "gene = KCNQ2, variant start: 62062719 variant end: 62062720sift: tolerated(0.08) polyphen: benign(0.012)\n",
      "gene = KCNQ2, variant start: 62065099 variant end: 62065100sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62065108 variant end: 62065109sift: tolerated_low_confidence(1) polyphen: unknown(0)\n",
      "gene = KCNQ2, variant start: 62065115 variant end: 62065116sift: deleterious_low_confidence(0.01) polyphen: unknown(0)\n",
      "gene = KCNQ2, variant start: 62065115 variant end: 62065116sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62065119 variant end: 62065120sift: tolerated_low_confidence(0.11) polyphen: unknown(0)\n",
      "gene = KCNQ2, variant start: 62065123 variant end: 62065124sift: tolerated_low_confidence(0.85) polyphen: unknown(0)\n",
      "gene = KCNQ2, variant start: 62065124 variant end: 62065125sift: tolerated_low_confidence(0.79) polyphen: unknown(0)\n",
      "gene = KCNQ2, variant start: 62065132 variant end: 62065133sift: tolerated_low_confidence(1) polyphen: unknown(0)\n",
      "gene = KCNQ2, variant start: 62065137 variant end: 62065138sift: tolerated_low_confidence(0.07) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62065140 variant end: 62065141sift: deleterious_low_confidence(0.03) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62065149 variant end: 62065150sift: tolerated_low_confidence(0.57) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62065150 variant end: 62065151sift: deleterious_low_confidence(0.01) polyphen: benign(0.003)\n",
      "gene = KCNQ2, variant start: 62065150 variant end: 62065151sift: tolerated_low_confidence(0.33) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62065153 variant end: 62065154sift: deleterious_low_confidence(0.04) polyphen: benign(0.001)\n",
      "gene = KCNQ2, variant start: 62065153 variant end: 62065154sift: tolerated_low_confidence(0.57) polyphen: benign(0)\n",
      "gene = KCNQ2, variant start: 62065156 variant end: 62065157sift: deleterious(0.02) polyphen: benign(0.217)\n",
      "gene = KCNQ2, variant start: 62065164 variant end: 62065165sift: deleterious(0.01) polyphen: probably_damaging(0.997)\n",
      "gene = KCNQ2, variant start: 62065167 variant end: 62065168sift: tolerated(0.08) polyphen: probably_damaging(0.974)\n",
      "gene = KCNQ2, variant start: 62065170 variant end: 62065171sift: deleterious(0.03) polyphen: probably_damaging(0.997)\n",
      "gene = KCNQ2, variant start: 62065176 variant end: 62065177sift: tolerated(0.78) polyphen: benign(0.005)\n",
      "gene = KCNQ2, variant start: 62065176 variant end: 62065177sift: tolerated(0.06) polyphen: benign(0.116)\n",
      "gene = KCNQ2, variant start: 62065177 variant end: 62065178sift: tolerated(0.78) polyphen: benign(0.005)\n",
      "gene = KCNQ2, variant start: 62065224 variant end: 62065225sift: deleterious(0) polyphen: probably_damaging(0.949)\n",
      "gene = KCNQ2, variant start: 62065230 variant end: 62065231sift: deleterious(0.01) polyphen: probably_damaging(0.985)\n",
      "gene = KCNQ2, variant start: 62065255 variant end: 62065256sift: deleterious(0) polyphen: benign(0.216)\n",
      "gene = KCNQ2, variant start: 62070002 variant end: 62070003sift: deleterious(0.02) polyphen: benign(0.417)\n",
      "gene = KCNQ2, variant start: 62070952 variant end: 62070953sift: deleterious(0.01) polyphen: probably_damaging(0.937)\n",
      "gene = KCNQ2, variant start: 62070991 variant end: 62070992sift: tolerated(0.51) polyphen: benign(0.004)\n",
      "gene = KCNQ2, variant start: 62071045 variant end: 62071046sift: tolerated(0.18) polyphen: benign(0.433)\n",
      "gene = KCNQ2, variant start: 62073767 variant end: 62073768sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62073799 variant end: 62073800sift: deleterious(0.02) polyphen: possibly_damaging(0.635)\n",
      "gene = KCNQ2, variant start: 62073862 variant end: 62073863sift: deleterious(0) polyphen: benign(0.007)\n",
      "gene = KCNQ2, variant start: 62076114 variant end: 62076115sift: deleterious(0.04) polyphen: probably_damaging(0.926)\n",
      "gene = KCNQ2, variant start: 62076130 variant end: 62076131sift: tolerated(0.33) polyphen: benign(0.023)\n",
      "gene = KCNQ2, variant start: 62076163 variant end: 62076164sift: tolerated(0.99) polyphen: benign(0.007)\n",
      "gene = KCNQ2, variant start: 62076172 variant end: 62076173sift: tolerated(0.43) polyphen: benign(0.12)\n",
      "gene = KCNQ2, variant start: 62076183 variant end: 62076184sift: deleterious(0) polyphen: possibly_damaging(0.812)\n",
      "gene = KCNQ2, variant start: 62076604 variant end: 62076605sift: deleterious(0) polyphen: probably_damaging(1)\n",
      "gene = KCNQ2, variant start: 62076610 variant end: 62076611sift: deleterious(0.01) polyphen: probably_damaging(0.998)\n",
      "gene = KCNQ2, variant start: 62076619 variant end: 62076620sift: tolerated(0.87) polyphen: benign(0.097)\n",
      "gene = KCNQ2, variant start: 62076640 variant end: 62076641sift: deleterious(0.01) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62076641 variant end: 62076642sift: deleterious(0) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62076655 variant end: 62076656sift: deleterious(0) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62076677 variant end: 62076678sift: tolerated(0.45) polyphen: probably_damaging(0.998)\n",
      "gene = KCNQ2, variant start: 62076677 variant end: 62076678sift: deleterious(0) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62076689 variant end: 62076690sift: tolerated(0.2) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62076701 variant end: 62076702sift: deleterious(0.01) polyphen: probably_damaging(0.998)\n",
      "gene = KCNQ2, variant start: 62076701 variant end: 62076702sift: deleterious(0.01) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62078104 variant end: 62078105sift: tolerated(0.28) polyphen: benign(0.024)\n",
      "gene = KCNQ2, variant start: 62078112 variant end: 62078113sift: tolerated(0.45) polyphen: benign(0.358)\n",
      "gene = KCNQ2, variant start: 62078118 variant end: 62078119sift: tolerated(0.42) polyphen: benign(0.29)\n",
      "gene = KCNQ2, variant start: 62078135 variant end: 62078136sift: tolerated(0.08) polyphen: probably_damaging(0.984)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene = KCNQ2, variant start: 62078143 variant end: 62078144sift: deleterious(0.04) polyphen: benign(0.006)\n",
      "gene = KCNQ2, variant start: 62078164 variant end: 62078165sift: deleterious(0) polyphen: benign(0.403)\n",
      "gene = KCNQ2, variant start: 62078174 variant end: 62078175sift: tolerated(0.53) polyphen: benign(0.002)\n",
      "gene = KCNQ2, variant start: 62103602 variant end: 62103603sift: tolerated(0.11) polyphen: probably_damaging(0.99)\n",
      "gene = KCNQ2, variant start: 62103614 variant end: 62103615sift: tolerated(0.85) polyphen: benign(0.005)\n",
      "gene = KCNQ2, variant start: 62103623 variant end: 62103624sift: tolerated(0.23) polyphen: benign(0.286)\n",
      "gene = KCNQ2, variant start: 62103626 variant end: 62103627sift: tolerated(0.77) polyphen: benign(0.07)\n",
      "gene = KCNQ2, variant start: 62103631 variant end: 62103632sift: tolerated(0.36) polyphen: benign(0.198)\n",
      "gene = KCNQ2, variant start: 62103656 variant end: 62103657sift: deleterious(0.02) polyphen: possibly_damaging(0.771)\n",
      "gene = KCNQ2, variant start: 62103667 variant end: 62103668sift: deleterious(0) polyphen: probably_damaging(0.987)\n",
      "gene = KCNQ2, variant start: 62103688 variant end: 62103689sift: tolerated(0.41) polyphen: possibly_damaging(0.478)\n",
      "gene = KCNQ2, variant start: 62103728 variant end: 62103729sift: tolerated(0.33) polyphen: benign(0.445)\n",
      "gene = KCNQ2, variant start: 62103731 variant end: 62103732sift: deleterious(0.03) polyphen: benign(0.007)\n",
      "gene = KCNQ2, variant start: 62103731 variant end: 62103732sift: tolerated(0.19) polyphen: benign(0.004)\n",
      "gene = KCNQ2, variant start: 62103739 variant end: 62103740sift: tolerated(0.11) polyphen: probably_damaging(0.999)\n",
      "gene = KCNQ2, variant start: 62103758 variant end: 62103762sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62103770 variant end: 62103771sift: tolerated_low_confidence(0.1) polyphen: benign(0.163)\n",
      "gene = KCNQ2, variant start: 62103773 variant end: 62103774sift: tolerated_low_confidence(0.27) polyphen: benign(0.007)\n",
      "gene = KCNQ2, variant start: 62103782 variant end: 62103783sift: tolerated_low_confidence(0.11) polyphen: possibly_damaging(0.601)\n",
      "gene = KCNQ2, variant start: 62103783 variant end: 62103784sift:  polyphen: \n",
      "gene = KCNQ2, variant start: 62103794 variant end: 62103795sift: deleterious_low_confidence(0.03) polyphen: possibly_damaging(0.609)\n",
      "gene = KCNQ2, variant start: 62103809 variant end: 62103810sift: tolerated_low_confidence(0.13) polyphen: benign(0.079)\n"
     ]
    }
   ],
   "source": [
    "## Read through gnomad EXOME vcf variant file\n",
    "exome_local_by_gene = processVCF(vcf_path=exome_vcf_path, gff_path=gff_path, sift_polyphen_filter=sift_polyphen_filter, sift_cutoff=sift_cutoff, polyphen_cutoff=polyphen_cutoff, type=\"exome sequencing\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter variants based on coverage cutoffs \n",
    "See [CCR github repo](https://github.com/quinlan-lab/ccr/blob/master/getfiles.sh) for information on the original files used. To pass the coverage filter, greater than **50% of the samples have to have > 10X coverage at the variant's position.** \n",
    "\n",
    "\n",
    "```\n",
    "Exome Sequencing: \n",
    "cd ~/git/quinlan_lab_rotation/reference/coverage/exome\n",
    "for i in {1..22} X Y; do wget -P data https://storage.googleapis.com/gnomad-public/release/2.0.1/coverage/exomes/gnomad.exomes.r2.0.1.chr$i.coverage.txt.gz; done\n",
    "for i in {1..22} X Y; do wget -P data https://storage.googleapis.com/gnomad-public/release/2.0.1/coverage/exomes/gnomad.exomes.r2.0.1.chr$i.coverage.txt.gz.tbi; done\n",
    "mv data/* .\n",
    "rm -rf data  \n",
    "\n",
    "Whole Genome Sequencing: \n",
    "cd ~/git/quinlan_lab_rotation/reference/coverage/wgs/\n",
    "for i in {1..22} X Y; do wget -P data https://storage.googleapis.com/gnomad-public/release/2.0.1/coverage/genomes/gnomad.genomes.r2.0.1.chr$i.coverage.txt.gz; done\n",
    "for i in {1..22} X Y; do wget -P data https://storage.googleapis.com/gnomad-public/release/2.0.1/coverage/genomes/gnomad.genomes.r2.0.1.chr$i.coverage.txt.gz.tbi; done\n",
    "mv data/* .\n",
    "rm -rf data \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to apply coverage filters \n",
    "def coverage_filter(data, coverage_dir): \n",
    "    filter_variants = defaultdict(list)\n",
    "    prev_chr = \"temp\"\n",
    "    for key in data.keys(): \n",
    "        subset_variants = data[key]\n",
    "        coverage_variants_per_gene = []\n",
    "        \n",
    "        ## Get the chromosome \n",
    "        chr = subset_variants[0][\"chr\"]\n",
    "  \n",
    "        ## Read in the coverage data\n",
    "        if prev_chr != chr: \n",
    "            coverage_file = coverage_dir + str(chr) + \".coverage.txt.gz\"\n",
    "            coverage_df = pd.read_csv(coverage_file, sep=\"\\t\")\n",
    "            print(\"reading in \" + coverage_file)\n",
    "#         else: \n",
    "#             coverage_df = coverage_df\n",
    "        \n",
    "        ## Go through each variant within each gene and filter based on coverage\n",
    "        for variant in subset_variants: \n",
    "            chr_start = variant[\"chr_start\"]\n",
    "            chr_end = variant[\"chr_stop\"]\n",
    "                \n",
    "            ## Remove sites that do not meet the coverage cutoff \n",
    "            variant_coverage = coverage_df[(coverage_df.pos >= chr_start) & \n",
    "                                          (coverage_df.pos <= chr_end)]\n",
    "            coverage_10X = variant_coverage['10'].tolist()\n",
    "            if not all(i >= .50 for i in coverage_10X): continue\n",
    "            \n",
    "            ## Append variant to the coverage_variants_per_gene dataset\n",
    "            coverage_variants_per_gene.append(variant)\n",
    "        \n",
    "        ## Add variants to final dataset \n",
    "        filter_variants[key] = coverage_variants_per_gene\n",
    "        \n",
    "        prev_chr = chr\n",
    "    return(filter_variants)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov_filter_df_exome = coverage_filter(data=exome_local_by_gene, coverage_dir=exome_coverage_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov_filter_df_wgs = coverage_filter(data=wgs_local_by_gene, coverage_dir=wgs_coverage_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the VAF percentiles to determine appropriate cutoffs\n",
    "* Use all variants that are above the n-th percentile of VAFs \n",
    "* Get the VAF percentiles from exome and whole genome variants separately "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Function to obtain list of vaf values from exome and wgs variants \n",
    "def get_vaf_values(vcf): \n",
    "    vaf_list = []\n",
    "    for key in vcf.keys(): \n",
    "        for variant in vcf[key]: \n",
    "            vaf_list.append(variant[\"vaf\"])\n",
    "    return(vaf_list)\n",
    "\n",
    "## Function to apply vaf filter on exome and wgs variants \n",
    "def apply_vaf_filter(vcf, cutoff):  \n",
    "    vaf_all_genes = defaultdict(list)\n",
    "    for key in vcf.keys():\n",
    "        vaf_by_gene = []\n",
    "        for variant in vcf[key]: \n",
    "            if variant[\"vaf\"] < cutoff: \n",
    "                continue\n",
    "            vaf_by_gene.append(variant)\n",
    "        vaf_all_genes[key] = vaf_by_gene\n",
    "    return(vaf_all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not vaf_filter: \n",
    "    exome_variants = cov_filter_df_exome\n",
    "    wgs_variants = cov_filter_df_wgs\n",
    "if vaf_filter: \n",
    "    ## Get the VAF values from exome and wgs samples\n",
    "    exome_vaf = get_vaf_values(vcf=cov_filter_df_exome)\n",
    "    exome_vaf = np.array(exome_vaf)\n",
    "    wgs_vaf = get_vaf_values(vcf=cov_filter_df_wgs)   \n",
    "    wgs_vaf = np.array(wgs_vaf)\n",
    "    \n",
    "    ## Get the numpy percentiles of vaf values \n",
    "    exome_vaf_cutoff = np.percentile(exome_vaf, vaf_cutoff)\n",
    "    wgs_vaf_cutoff = np.percentile(wgs_vaf, vaf_cutoff)\n",
    "    \n",
    "    ## Apply vaf filter on exome data \n",
    "    exome_variants = apply_vaf_filter(vcf=cov_filter_df_exome, cutoff=exome_vaf_cutoff)\n",
    "    \n",
    "    ## Apply vaf filter on wgs data \n",
    "    wgs_variants = apply_vaf_filter(vcf=cov_filter_df_wgs, cutoff=wgs_vaf_cutoff)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter variants based on segmental duplication and self-chains \n",
    "See [CCR github repo](https://github.com/quinlan-lab/ccr/blob/master/getfiles.sh) for information on the original files used. Creating the bed files requires the process (or a variation of the process) described below: \n",
    "\n",
    "```\n",
    "Segmental Duplications: \n",
    "cd ~/git/quinlan_lab_rotation/reference/segmental_duplications\n",
    "wget ftp://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/genomicSuperDups.txt.gz\n",
    "zcat < genomicSuperDups.txt.gz | cut -f 2-4 | sed 's/^chr//g' | sort -k1,1 -k2,2n > segmental_duplication_file.bed\n",
    "## zcat < genomicSuperDups.txt.gz | cut -f 2-4 | sed 's/^chr//g' | sort -k1,1 -k2,2n | bedtools merge | bgzip -c > data/segmental.bed.gz; cd data; tabix -f segmental.bed.gz\n",
    "\n",
    "Self Chains: \n",
    "cd ~/git/quinlan_lab_rotation/reference/self_chains\n",
    "wget ftp://hgdownload.cse.ucsc.edu/goldenpath/hg19/database/chainSelf.txt.gz\n",
    "gunzip chainSelf.txt.gz\n",
    "python get-chain.py | sed 's/^chr//g' | sort -k1,1 -k2,2n > selfchains_final.bed\n",
    "## python get-chain.py | sed 's/^chr//g' | sort -k1,1 -k2,2n | bedtools merge | bgzip -c > data/self-chains.id90.bed.gz; cd data; tabix -f self-chains.id90.bed.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in the segmental duplication and self-chain data \n",
    "segdup_data = pd.read_csv(\"/Users/jasonkunisaki/git/quinlan_lab_rotation/reference/segmental_duplications/segmental_duplication_file.bed\", sep='\\t', header=None) \n",
    "selfchain_data = pd.read_csv(\"/Users/jasonkunisaki/git/quinlan_lab_rotation/reference/self_chains/selfchains_final.bed\", sep='\\t')\n",
    "remove_variants = selfchain_data\n",
    "remove_variants.append(segdup_data)\n",
    "remove_variants.columns = [\"chr\", \"start\", \"end\"]\n",
    "\n",
    "## Function to apply segmental duplication and self-chain filters\n",
    "def remove_segdup_selfchains(data, remove_variants=remove_variants):\n",
    "    filter_variants = defaultdict(list)\n",
    "    for key in data.keys(): \n",
    "        subset_variants = data[key] ## Subset variants by gene \n",
    "        variants_non_segdup_selfchain_per_gene = []\n",
    "        for variant in subset_variants: \n",
    "            chromosome = variant[\"chr\"]\n",
    "            chr_start = variant[\"chr_start\"]\n",
    "            chr_end = variant[\"chr_stop\"]\n",
    "\n",
    "            ## Subset remove_variants df based on chr\n",
    "            temp_rm_variants = remove_variants[remove_variants.chr==chromosome]\n",
    "            temp_rm_variants = temp_rm_variants[(((temp_rm_variants.start <= chr_start) & \n",
    "                                                (temp_rm_variants.end >= chr_start)) | ## if the variant's start is within the bad region \n",
    "                                               ((temp_rm_variants.start <= chr_end) & \n",
    "                                                (temp_rm_variants.end >= chr_end)) | ## if the variant's end is within the bad region\n",
    "                                                ((temp_rm_variants.start >= chr_start) &\n",
    "                                                temp_rm_variants.end <= chr_end))] ## if the bad region is within the variant's boundaries\n",
    "\n",
    "            ## Remove variants if the chr_start/end positions are within the remove_variants windows\n",
    "            if len(temp_rm_variants) > 0: \n",
    "                #print(variant)\n",
    "                continue \n",
    "            if len(temp_rm_variants) == 0: \n",
    "                variants_non_segdup_selfchain_per_gene.append(variant)\n",
    "        \n",
    "        filter_variants[key] = variants_non_segdup_selfchain_per_gene\n",
    "    return(filter_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_segdup_selfchain_exome = remove_segdup_selfchains(data=exome_variants, remove_variants=remove_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_segdup_selfchain_wgs = remove_segdup_selfchains(data=wgs_variants, remove_variants=remove_variants)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combine the exome and whole genome variants that passed the above filters: \n",
    "* VAF cutoff \n",
    "* SIFT/PolyPhen scores \n",
    "* Remove segmental duplications and self-chains \n",
    "* Coverage "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Combine wgs and exome variant data. Duplicate variants will be accounted for below. \n",
    "combined_wgs_exome_variants = defaultdict(list)\n",
    "for key in no_segdup_selfchain_wgs.keys(): \n",
    "    if len(no_segdup_selfchain_exome[key]) == 0: \n",
    "        combined_wgs_exome_variants[key] = no_segdup_selfchain_wgs[key]\n",
    "    else: \n",
    "        combined_wgs_exome_variants[key] = no_segdup_selfchain_exome[key] + no_segdup_selfchain_wgs[key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_wgs_exome_variants[test_key]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform quality check on pathogenic variants. Remove genes with no variants after applying filters from above. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_variant_vcf = defaultdict(list)\n",
    "## Check if there is at least one variant in each gene \n",
    "for key in combined_wgs_exome_variants.keys():\n",
    "    if len(combined_wgs_exome_variants[key]) > 0:\n",
    "        final_variant_vcf[key] = combined_wgs_exome_variants[key]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fix overlapping variants within each gene + apply SIFT/PolyPhen and VAF filters\n",
    "| Position |   V1   |   V2   |   V3   |   V4   |   V5   |   V6   |   V7   |   V8   |   V9   |  V10   |  V11   |\n",
    "| :------- | :----: | :----: | :----: | :----: | :----: | :----: | :----: | :----: | :----: | :----: | :----: |\n",
    "| Start (0-based)    | 126244 | 126244 | 126253 | 126255 | 126258 | 126258 | 126260 | 126264 | 126264 | 126265 | **126279** |\n",
    "| End (1-based)      | 126245 | 126254 | 126254 | 126256 | 126259 | 126262 | **126279** | 126265 | 126265 | 126266 | 126280 |\n",
    "\n",
    "### Combine variant positions for:\n",
    "+ adjacent bases: **V7 ends at 126279 (1-based) and V11 starts at 126279 (0-based). The variant windows are combined.**\n",
    "+ equivalent variants (in terms of position) \n",
    "+ overlapping variants\n",
    "\n",
    "### The determined variant range should encompass all of the above variants\n",
    "| Position | window 1 | window 2 | window 3 | \n",
    "| :------: | :------: | :------: | :------: |\n",
    "| Start    | 126244   | 126255   | 126258   |\n",
    "| End      | 126254   | 126256   | 126280   |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Inititalize the master list \n",
    "unique_variant_pos_all_genes = defaultdict(list)\n",
    "for key in final_variant_vcf.keys():\n",
    "    subset_variants = final_variant_vcf[key]\n",
    "    ## Get all of the start and stop variant positions within the gene\n",
    "    start_positions = [d[\"chr_start\"] for d in subset_variants]\n",
    "    end_positions = [d[\"chr_stop\"] for d in subset_variants]\n",
    "    exonic_end_positions = [d[\"stop\"] for d in subset_variants]\n",
    "\n",
    "    ## Go throouh each start/end position and check if there are overlapping variants \n",
    "    unique_variant_pos_per_gene = []\n",
    "    positions = []\n",
    "    for start, end in zip(start_positions, end_positions):\n",
    "        ## Get variants in which the start is within the variant's start/end\n",
    "        temp1 = list(filter(lambda subset_variants: \n",
    "               subset_variants[\"chr_start\"] <= start and subset_variants[\"chr_stop\"] >= start, subset_variants))\n",
    "        \n",
    "        ## Check to see if there are variants that overlap with maximum variant's end position from the above subset\n",
    "        new_end = max([d[\"chr_stop\"] for d in temp1])\n",
    "        temp2 = list(filter(lambda subset_variants:\n",
    "                   subset_variants[\"chr_start\"] <= new_end and subset_variants[\"chr_stop\"] >= new_end, subset_variants))\n",
    "        \n",
    "        ## Combine all variants that overlap with the intial variant(s) in question\n",
    "        for v in temp2: \n",
    "            if v not in temp1: \n",
    "                temp1.append(v)\n",
    "\n",
    "        ## Get the index at which an adjacent variant's start position overlaps with the end position of the adjusted window\n",
    "        ## Assign the maximum end pos of the adjacent variant as the window's stop position\n",
    "        while max(set([d[\"chr_stop\"] for d in temp1])) in start_positions and \\\n",
    "        max(set([d[\"chr_stop\"] for d in temp1])) not in positions: ## Make sure it does not pull a variant that was already accounted for in a window\n",
    "            d = defaultdict(list)\n",
    "            for index, e in enumerate(start_positions):\n",
    "                d[e].append(index)\n",
    "            \n",
    "            ## Get all the end positions of adjacent variants \n",
    "            end_index = d[max(set([d[\"chr_stop\"] for d in temp1]))]\n",
    "            \n",
    "            ## Replace the maximum of the adjusted window with the adjacent variant's stop\n",
    "            temp1[len(temp1)-1][\"stop\"] = max(list(np.array(exonic_end_positions)[end_index]))\n",
    "            temp1[len(temp1)-1][\"chr_stop\"] = max(list(np.array(end_positions)[end_index]))\n",
    "\n",
    "        ## If the variant is NOT the first variant \n",
    "        if len(positions) > 0: \n",
    "            ## Check to see if the variant's start and stop falls in a window already calculated \n",
    "            prev_start_positions = [d[\"final_start_positions\"] for d in positions] ## Get start positions\n",
    "            prev_end_positions = [d[\"final_end_positions\"] for d in positions] ## Get end positions\n",
    "            for s, e in zip(prev_start_positions, prev_end_positions):\n",
    "                ## If the variant' start or end falls within a previously calculated window, skip the variant \n",
    "                if (start >= s and start <= e) or (end >= s and start <= e):\n",
    "                    temp1 = []\n",
    "                    break\n",
    "\n",
    "        ## If there are overlapping variants\n",
    "        if len(temp1) > 1: \n",
    "            final_temp = {\n",
    "                \"gene\": key,\n",
    "                \"start\": min(set([d[\"start\"] for d in temp1])),\n",
    "                \"stop\": max(set([d[\"stop\"] for d in temp1])),\n",
    "                \"strand\": temp1[1][\"strand\"],\n",
    "                \"exon\": temp1[1][\"exon\"],\n",
    "                \"chr\": temp1[1][\"chr\"],\n",
    "                \"chr_start\": min(set([d[\"chr_start\"] for d in temp1])),\n",
    "                \"chr_stop\": max(set([d[\"chr_stop\"] for d in temp1])),\n",
    "                \"mutation\": \"overlapping mutations\",\n",
    "                \"vaf\": temp1[1][\"vaf\"],\n",
    "                \"SIFT\": temp1[1][\"SIFT\"],\n",
    "                \"PolyPhen\": temp1[1][\"PolyPhen\"]\n",
    "            }\n",
    "            unique_variant_pos_per_gene.append(final_temp)\n",
    "\n",
    "        ## If there are no overlapping variants \n",
    "        if len(temp1) == 1:\n",
    "            final_temp = temp1[0]\n",
    "            unique_variant_pos_per_gene.append(final_temp)\n",
    "\n",
    "        ## Get the complete start and stop positions for variants within gene\n",
    "        test = {\n",
    "        \"final_start_positions\": final_temp[\"chr_start\"], ## 0-based\n",
    "        \"final_end_positions\": final_temp[\"chr_stop\"] ## 1-based \n",
    "        } \n",
    "\n",
    "        ## Store the start and stop positions for each adjusted variant range\n",
    "        if len(temp1) > 0:\n",
    "            positions.append(test)\n",
    "    \n",
    "    ## Sort each variant within each gene based on the genomic start position\n",
    "    unique_variant_pos_per_gene = sorted(unique_variant_pos_per_gene, key = lambda i: i[\"chr_start\"])\n",
    "    unique_variant_pos_all_genes[key] = unique_variant_pos_per_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_variant_pos_all_genes[test_key]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Caclulate CCR window sizes and positions based on variant coordinates\n",
    "```\n",
    "   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  19  20    ## 1 base\n",
    " | T | T | G | A | T | C | G | A | T | A | G | C | T | A | C | C | G | A | T | C |\n",
    " 0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  19  20  ## 0 base \n",
    "```\n",
    "**Examples**\n",
    "+ **1 base b/w variants:** 2 *(1-based; variant_plus_one.start; 2-3 G)* - 1 *(0-based; variant.end; 0-1 T)* = 1\n",
    "+ **Multiple bases b/w variants**: 10 *(0-based; variant_plus_one.start; 10-11 G)* - 4 *(1-based; variant.end; 0-4 TTGA)* = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Store the genes as keys \n",
    "## Store the window sizes/variant positions in dictionaries  \n",
    "final_window_dataset = {}\n",
    "for key in unique_variant_pos_all_genes.keys(): \n",
    "    ## Subset dictionaries (i.e. list of variants) based on each key (i.e. each gene)\n",
    "    subset_variants = unique_variant_pos_all_genes[key]\n",
    "    \n",
    "    ## Go through the variants within each gene and measure the CCR window size \n",
    "    subset_window_list = []\n",
    "    \n",
    "    ## Go through each variant, get the variant pair, and then calculate window size \n",
    "    for variant, variant_plus_one in zip(subset_variants, subset_variants[1:]):\n",
    "        ## Calculate window position of the variant\n",
    "        variant1_window = {\"gene\": key,\n",
    "        \"chromosome\": \"chr\" + variant[\"chr\"],\n",
    "        \"variant1_start\": variant[\"start\"],\n",
    "        \"variant1_stop\": variant[\"stop\"],\n",
    "        \"variant2_start\": \"NA\",\n",
    "        \"variant2_stop\": \"NA\",\n",
    "        \"variant1_vaf\": variant[\"vaf\"],\n",
    "        \"variant2_vaf\": \"NA\",\n",
    "        \"variant1_mut\": variant[\"mutation\"],\n",
    "        \"variant2_mut\": \"NA\",\n",
    "        \"variant1_SIFT\": variant[\"SIFT\"],\n",
    "        \"variant1_PolyPhen\": variant[\"PolyPhen\"],\n",
    "        \"variant2_SIFT\": \"NA\",\n",
    "        \"variant2_PolyPhen\": \"NA\",\n",
    "        \"window_size\": 0,\n",
    "        \"window_start\": variant[\"chr_start\"],\n",
    "        \"window_stop\": variant[\"chr_stop\"],\n",
    "        \"varflag\": \"VARTRUE\"\n",
    "        }\n",
    "\n",
    "        ## Calculate window positions between variants\n",
    "        window_dict = {\"gene\": key,\n",
    "        \"chromosome\": \"chr\" + variant[\"chr\"],\n",
    "        \"variant1_start\": variant[\"start\"],\n",
    "        \"variant1_stop\": variant[\"stop\"],\n",
    "        \"variant2_start\": variant_plus_one[\"start\"],\n",
    "        \"variant2_stop\": variant_plus_one[\"stop\"],\n",
    "        \"variant1_vaf\": variant[\"vaf\"],\n",
    "        \"variant2_vaf\": variant_plus_one[\"vaf\"],\n",
    "        \"variant1_mut\": variant[\"mutation\"],\n",
    "        \"variant2_mut\": variant_plus_one[\"mutation\"],\n",
    "        \"variant1_SIFT\": \"NA\",\n",
    "        \"variant1_PolyPhen\": \"NA\",\n",
    "        \"variant2_SIFT\": \"NA\",\n",
    "        \"variant2_PolyPhen\": \"NA\",\n",
    "        \"window_size\": max(0, variant_plus_one[\"start\"] - variant[\"stop\"]),\n",
    "        \"window_start\": variant[\"chr_stop\"],\n",
    "        \"window_stop\": variant_plus_one[\"chr_start\"], \n",
    "        \"varflag\": \"VARFALSE\"\n",
    "        }\n",
    "\n",
    "        ## Append window for variant and adjacent variants\n",
    "        subset_window_list.append(variant1_window)\n",
    "        if window_dict[\"window_size\"] > 0 :\n",
    "            subset_window_list.append(window_dict)\n",
    "\n",
    "    ## Get the information for the last variant in the GOI \n",
    "    last_variant = subset_variants[len(subset_variants)-1]\n",
    "    last_variant = {\"gene\": key,\n",
    "        \"chromosome\": \"chr\" + last_variant[\"chr\"],\n",
    "        \"variant1_start\": last_variant[\"start\"],\n",
    "        \"variant1_stop\": last_variant[\"stop\"],\n",
    "        \"variant2_start\": \"NA\",\n",
    "        \"variant2_stop\": \"NA\",\n",
    "        \"variant1_vaf\": last_variant[\"vaf\"],\n",
    "        \"variant2_vaf\": \"NA\",\n",
    "        \"variant1_mut\": last_variant[\"mutation\"],\n",
    "        \"variant2_mut\": \"NA\",\n",
    "        \"variant1_SIFT\": last_variant[\"SIFT\"],\n",
    "        \"variant1_PolyPhen\": last_variant[\"PolyPhen\"],\n",
    "        \"variant2_SIFT\": \"NA\",\n",
    "        \"variant2_PolyPhen\": \"NA\",\n",
    "        \"window_size\": 0,\n",
    "        \"window_start\": last_variant[\"chr_start\"],\n",
    "        \"window_stop\": last_variant[\"chr_stop\"],\n",
    "        \"varflag\": \"VARTRUE\"\n",
    "        }\n",
    "    \n",
    "    ## Append window for the last variant \n",
    "    subset_window_list.append(last_variant)\n",
    "\n",
    "    ## Append each gene's window data to the associated gene in the final dataset\n",
    "    assert(key not in final_window_dataset)\n",
    "    final_window_dataset[key] = subset_window_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_window_dataset[test_key]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Obtain the percentiles for each CCR window \n",
    "## Calculate percentiles based on varflag == VARFALSE value "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define the 'windows' list  \n",
    "windows = []\n",
    "\n",
    "## Go through each gene and get the window sizes\n",
    "for key in final_window_dataset.keys():\n",
    "    subset_variants = final_window_dataset[key]\n",
    "    for variant in subset_variants:\n",
    "#         if variant[\"varflag\"] == \"VARFALSE\":\n",
    "#             windows.append(variant[\"window_size\"])\n",
    "          windows.append(variant[\"window_size\"])\n",
    "\n",
    "## Sort the 'windows' list\n",
    "windows.sort()\n",
    "\n",
    "## Convert windows to numpy array\n",
    "windows = np.array(windows)\n",
    "\n",
    "## Go through each of the variants and obtain the scores \n",
    "final_dataset = defaultdict(list)\n",
    "for key in final_window_dataset.keys():\n",
    "    ## Get the variants found within each gene\n",
    "    subset_variants = final_window_dataset[key]\n",
    "    \n",
    "    ## Define a percentiles list with each window's percentile + other information as the dictionary\n",
    "    percentiles_single_gene = []\n",
    "    for variant in subset_variants:\n",
    "        ## If the variant is a variant, assign percentile as 0\n",
    "        if variant[\"varflag\"] == \"VARTRUE\": \n",
    "            variant[\"percentile\"] = 0\n",
    "        \n",
    "        ## If the variant is a CCR window, assign percentile based on window values \n",
    "        if variant[\"varflag\"] == \"VARFALSE\":        \n",
    "            variant[\"percentile\"] = round(percentileofscore(windows, variant[\"window_size\"]), 2)\n",
    "        \n",
    "        ## Append the dictionary to the list of percentiles for each gene\n",
    "        percentiles_single_gene.append(variant)\n",
    "    \n",
    "    ## Appened all percentiles across all genes to the final dataset list \n",
    "    final_dataset[key] = percentiles_single_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_dataset[test_key]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the weighted percentiles \n",
    "Start at 100 (the most constrained region) and scale down proportionally for each subsequent region by the fraction of the length of protein-coding exome already covered by the preceding region. More information can be found [here](https://quinlan-lab.github.io/ccr/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Use panda dataframes to make BED files for PathoScore\n",
    "new_CCR = pd.DataFrame()\n",
    "for key in final_dataset.keys():\n",
    "    gene_pandaDf = pd.DataFrame(final_dataset[key], columns=[\"chromosome\", \"window_start\", \n",
    "        \"window_stop\", \"gene\", \n",
    "        \"window_size\", \"varflag\", \"variant1_vaf\", \"variant1_SIFT\", \"variant1_PolyPhen\", \"percentile\"])\n",
    "    new_CCR = pd.concat([new_CCR, gene_pandaDf])\n",
    "\n",
    "## Change the column names of the combined panda dataframes and remove window_size column\n",
    "new_CCR.columns = [\"#chrom\", \"start\", \"end\", \"gene\", \"window_size\", \"varflag\",\n",
    "                   \"vaf\", \"SIFT\", \"PolyPhen\", \"residual_pct\"]\n",
    "\n",
    "## Sort (descending) the panda dataframe by residual percentile \n",
    "new_CCR = new_CCR.sort_values(by = [\"residual_pct\", \"#chrom\", \"start\"], ascending = False)\n",
    "\n",
    "## Write the output file and read it back in to remove duplicate index values in pandas dataset \n",
    "output_file = pathoscore_name + \".bed\" \n",
    "new_CCR.to_csv(output_file, sep=\"\\t\", index=False)\n",
    "new_CCR = pd.read_csv(pathoscore_name + \".bed\", sep=\"\\t\")\n",
    "\n",
    "## Print the head of new_CCR dataset\n",
    "new_CCR.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get the total length covered by CCRs\n",
    "totlength = sum(new_CCR[\"window_size\"].tolist())\n",
    "\n",
    "## Go throuh each row and determine the weighted percentile \n",
    "pct = 100.0\n",
    "opct = new_CCR.loc[0, \"residual_pct\"]\n",
    "weighted = [] ## Initialize the weighted percentile list \n",
    "for r in new_CCR.itertuples():\n",
    "    ## Do not perform this calculation on variants\n",
    "    if r[6] != \"VARTRUE\":  \n",
    "        regionlength = int(r[5]) ## Get the CCR size \n",
    "        residual_pct = r[len(r)-1] ## Get the residual CCR percentile \n",
    "        \n",
    "        ## If the percentile is not equal to the previous CCR's percentile\n",
    "        if r[len(r)-1] != opct: \n",
    "            pct -= regionlength/totlength*100 ## Perform the adjustment calculation \n",
    "            opct = residual_pct ## Assign new CCR percentile\n",
    "                             \n",
    "        ## Add weighted percentile to the weighted list \n",
    "        weighted.append(pct)\n",
    "    else: ## Assign 0th percentile to variants \n",
    "        weighted.append(0)\n",
    "        \n",
    "## TODO: check to see if these percentiles are the final weighted percentiles \n",
    "X_train = np.array(weighted).reshape(len(weighted), 1)\n",
    "min_max_scaler = preprocessing.MinMaxScaler(feature_range=(0,100))\n",
    "final_pctile = min_max_scaler.fit_transform(X_train)\n",
    "final_pctile_df = pd.DataFrame(final_pctile)\n",
    "\n",
    "## Add the weighted CCR values to the panda datase \n",
    "weighted_CCR = pd.concat([new_CCR, final_pctile_df], axis=1)\n",
    "weighted_CCR.columns = [\"#chrom\", \"start\", \"end\", \"gene\", \"window_size\", \"varflag\",\n",
    "                   \"vaf\", \"SIFT\", \"PolyPhen\", \"residual_pct\", \"ccr_pct\"]\n",
    "\n",
    "weighted_CCR.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a BED file to run in PathoScore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sort by position\n",
    "weighted_CCR = weighted_CCR.sort_values(by=[\"#chrom\", \"start\", \"end\"])\n",
    "\n",
    "## Write the output file and read it back in to remove duplicate index values in pandas dataset \n",
    "output_file = pathoscore_name + \".bed\" \n",
    "weighted_CCR.to_csv(output_file, sep=\"\\t\", index=False) ## Write the bed file \n",
    "weighted_CCR.to_csv((pathoscore_name + \".txt\"), sep=\"\\t\", index=False) ## Write the text file "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate the script needed to run the PathoScore function on all bed files present in the directory \n",
    "\n",
    "Step 1: create the scripts needed to run PathoScore \n",
    "\n",
    "Step 2: bgzip and tabix all of the bed files present in the directory (TODO)\n",
    "\n",
    "Step 3: annotate benign and pathogenic variants from all bed files \n",
    "\n",
    "Step 4: run the PathoScore evaluate function on annotated benign and pathogenic files \n",
    "\n",
    "## Write the PathoScore scripts based on the bed files present in the current working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_files = os.listdir(\".\")\n",
    "pattern = \"*.bed\"\n",
    "## Create a list of bed files there associated column numbers \n",
    "bedfile_colnum_list = []\n",
    "for file in list_of_files: \n",
    "    if fnmatch.fnmatch(file, pattern):\n",
    "        data = pd.read_csv(file, sep='\\t')\n",
    "        colnames = list(data.columns)\n",
    "        bedfile_colnum_list.append({\n",
    "            \"filename\": \"~/git/quinlan_lab_rotation/\" + file + \".gz\",\n",
    "            \"filename_no_suffix\": file.split(\".bed\")[0],\n",
    "            \"colnumber\": colnames.index(\"ccr_pct\") + 1\n",
    "        })\n",
    "\n",
    "## Use the bedfile names and column numbers to generate pathoscore scripts\n",
    "score_script = []\n",
    "evaluate_script = []\n",
    "for i in range(len(bedfile_colnum_list)): \n",
    "    filename = bedfile_colnum_list[i][\"filename\"]\n",
    "    filename_no_suffix = bedfile_colnum_list[i][\"filename_no_suffix\"]\n",
    "    colnum = str(bedfile_colnum_list[i][\"colnumber\"])\n",
    "    score_script.append(\"--scores \" + filename + \":\" + filename_no_suffix + \":\" + colnum + \":max\")\n",
    "    evaluate_script.append(\"-s \" + filename_no_suffix)\n",
    "separator = \" \"\n",
    "benign_score_script = \"python pathoscore.py annotate --prefix benign \" + separator.join(score_script) + \" ~/git/quinlan_lab_rotation/pathoscore/truth-sets/GRCh37/clinvar/clinvar-benign.20170905.vcf.gz\"\n",
    "pathogenic_score_script = \"python pathoscore.py annotate --prefix pathogenic \" + separator.join(score_script) + \" ~/git/quinlan_lab_rotation/pathoscore/truth-sets/GRCh37/clinvar/clinvar-pathogenic.20170905.vcf.gz\"\n",
    "pathoscore_evaluate_script = \"python pathoscore.py evaluate \" + separator.join(evaluate_script) + \" pathogenic.vcf.gz benign.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(benign_score_script)\n",
    "print(pathogenic_score_script)\n",
    "print(pathoscore_evaluate_script)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_object = open(r\"pathoscore_scripts.sh\", \"w+\")\n",
    "file_object.write(\"%s\\n%s\\n%s\\n\" % (benign_score_script, pathogenic_score_script, pathoscore_evaluate_script))\n",
    "file_object.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bgzip all of the bed files in the directory "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "bash /Users/jasonkunisaki/git/quinlan_lab_rotation/scripts/bgzip.sh /Users/jasonkunisaki/git/quinlan_lab_rotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run the PathoScore funtions that were generated above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "bash /Users/jasonkunisaki/git/quinlan_lab_rotation/pathoscore_scripts.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Move PathoScore output files to the appropriate output directory "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "bash /Users/jasonkunisaki/git/quinlan_lab_rotation/scripts/move_files.sh /Users/jasonkunisaki/git/quinlan_lab_rotation/pathoscore_results/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Print ROC curve from the PathoScore results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SVG('/Users/jasonkunisaki/git/quinlan_lab_rotation/pathoscore_results/pathoscore.roc.svg') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
