{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in the VCF file \n",
    "vcf = cyvcf2.VCF(vcf_path)\n",
    "\n",
    "## Get the VEP description fields into a 'list'\n",
    "kcsq = vcf[\"vep\"][\"Description\"].split(\":\")[1].strip(' \"').split(\"|\")\n",
    "\n",
    "## Read in the gff transcript file and subset calls from chr20\n",
    "transcripts = exter.read_gff(gff_path)\n",
    "transcripts20 = transcripts[\"20\"] # get chr20\n",
    "\n",
    "## Make an empty dictionary \n",
    "local_by_gene = defaultdict(list)\n",
    "\n",
    "## For each variant in the VCF file: \n",
    "for variant in vcf:\n",
    "    ## Remove variants that have \"None, PASS, SEGDUP, or LCR\" as the filter result\n",
    "    ## NOTE: 'continue' will remove variants without these FILTER results\n",
    "    if not (variant.FILTER is None or variant.FILTER in [\"PASS\", \"SEGDUP\", \"LCR\"]):\n",
    "            continue\n",
    "    ## Get the variant 'INFO' field from the VCF file for each variant\n",
    "    info = variant.INFO\n",
    "    ## Set the functional consequence of the variant to 'False'\n",
    "    any_functional = False\n",
    "    ## Merge the kcsq description keys with \n",
    "    csqs = [dict(zip(kcsq, c.split(\"|\"))) for c in info['vep'].split(\",\")]\n",
    "\n",
    "    ## For all variant/consequences that are within protein_coding regions\n",
    "    for csq in (c for c in csqs if c['BIOTYPE'] == 'protein_coding'):\n",
    "        ## NOTE: 'continue' will ignore variants that are intronic \n",
    "        if csq['Feature'] == '' or csq['EXON'] == '': \n",
    "            continue #or csq['cDNA_position'] == '': continue\n",
    "        ## Check to see if the variant's consequence is functional (based on the mutation types listed above)\n",
    "        ## NOTE: 'continue' will ignore variants that are not functional\n",
    "        if not isfunctional(csq): \n",
    "            continue\n",
    "        ## Assigns True to the variant as functional if it is NOT intronic and is deemed as functional by the 'isfunctional' function\n",
    "        any_functional = True\n",
    "\n",
    "        ## Get the exonic positions of the variants\n",
    "        local = transcripts20.localize(variant.start, variant.end)\n",
    "\n",
    "        ## Get the VAF of the variant\n",
    "        vaf = variant.INFO.get('AF')\n",
    "\n",
    "        ## Get the SIFT/PolyPhen results (IS THIS RIGHT)\n",
    "        SIFT = csq[\"SIFT\"]\n",
    "        PolyPhen = csq[\"PolyPhen\"]\n",
    "                \n",
    "        ## Get the reference allele \n",
    "        ref_allele = variant.REF\n",
    "\n",
    "        ## Get string for the mutation \n",
    "        mut = []\n",
    "        for alt in variant.ALT: \n",
    "            mut.append(ref_allele + \">\" + alt) \n",
    "        \n",
    "        ## Remove variants that do not meet a specific VAF cutoff \n",
    "        # if vaf < 0.0000001: continue\n",
    "        \n",
    "        ## Perform SIFT and Polyphen filters to remove the least delterious/damaging variant(s)\n",
    "        ## Go through each pair of scores and check if the varaint has SIFT and PolyPhen scores\n",
    "        sift_score = sift_cutoff + 0.1\n",
    "        polyphen_score = polyphen_cutoff - 0.1\n",
    "        if sift_polyphen_filter == True: \n",
    "            if re.findall(\"deleterious\", SIFT):\n",
    "                sift_score = float(re.split('\\)', re.split('\\(', SIFT)[1])[0])\n",
    "            if re.findall(\"damaging\", PolyPhen):\n",
    "                polyphen_score = float(re.split('\\)', re.split('\\(', PolyPhen)[1])[0])\n",
    "            ## SIFT score has to be below the sift_cutoff to pass \n",
    "            ## PolyPhen score has to be above the polyphen_cutoff to pass \n",
    "            if not sift_score < sift_cutoff and polyphen_score > polyphen_cutoff:\n",
    "                continue\n",
    "\n",
    "        ## Remove variants that are found in duplicated segments \n",
    "        #if variant.INFO.get('segdup') == True or variant.INFO.get('lcr'): continue\n",
    "\n",
    "        ## Get the chromosome \n",
    "        chr = variant.CHROM\n",
    "\n",
    "        for l in local:\n",
    "            l[\"chr\"] = chr\n",
    "            l[\"chr_start\"] = variant.start\n",
    "            l[\"chr_stop\"] = variant.end\n",
    "            l[\"mutation\"] = mut\n",
    "            l[\"vaf\"] = vaf\n",
    "            l[\"SIFT\"] = sift_score\n",
    "            l[\"PolyPhen\"] = polyphen_score\n",
    "            local_by_gene[l[\"gene\"]].append(l)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "local_by_gene[test_key]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
